File: 2 - Sorting and Order Statistics/2.1 - Heap Sort/Code/heap_sort.py
"""
heap_sort.py

"""

from .build_heap import BuildHeap


class HeapSort(BuildHeap):
    """Heap Sort Class"""

    def __init__(self, heap):
        super().__init__(heap)
        self.n = self.heap_size - 1

    def heap_sort(self):
        """Heap Sort Algorithm

        Time complexity: O(n log n)
        Space complexity: O(1)
        """
        self.build_max_heap()

        for i in range(self.n, 0, -1):
            self.heap[0], self.heap[i] = self.heap[i], self.heap[0]
            self.heap_size -= 1
            self.max_heapify(0)

File: 2 - Sorting and Order Statistics/2.1 - Heap Sort/Code/HeapSort.js
/**
 * HeapSort.js
 *
 */

const BuildHeap = require("./BuildHeap");

class HeapSort extends BuildHeap {
    constructor(heap) {
        super(heap);
        this.n = this.heapSize - 1;
    }

    heapSort() {
        this.buildMaxHeap();

        for (let i = this.n; i > 0; i--) {
            let temp = this.heap[0];
            this.heap[0] = this.heap[i];
            this.heap[i] = temp;
            this.heapSize -= 1;
            this.maxHeapify(0);
        }
    }
}

module.exports = HeapSort;
